### ISSUE TITLE
Squash commits before merge, but only *after* review?

### ISSUE THREAD

**Issue by DanVanAtta** [triplea-game/triplea#145]
Having a sequence of concise commits that are individually reviewable makes for an easier review IMO. It does not on the other hand create much of a nice history, and non-linear history problems can make things even more challenging.

So on the one hand, we want many commits for review purposes, but for concise history, a linear history, ideally one commit per merge (or per branch, or per PR). The linear history is an important one, non-linear history problems are a PITA.

How about this:
- When submitting a feature branch for PR, do not squash the check point commits
- Reviewers will check things out do some testing
- Reviewers will request changes as needed, contributor will make small commits with feedback changes
- Reviewers will give okay when things look good
- submitter squashes the commits together into just 1 or 2 single commits (large feature orientated commits) and pushes (which will trigger a travis build re-running the tests for a last time before merge)
- reviewers verify the final commit message, take a last look and merge the PR
  - then ideally master is tagged with an increment version number
  - this will trigger a travis run which will run tests (this is an integration test now that the code has been integrated to everything else that is on master. If the branch was recently based on master then this may be redundant, practically speaking that is not always the case)
  - artifacts are (then automatically) made and pushed, which allows then for further regression testing

**Comment by veqryn** [comment]
I guess this is fine, so long as we rebase before any squashing...
Can you give some example commands for the sequence of events?

**Comment by DanVanAtta** [comment]
This is probably more of a process thing. Question is *when* commits are squashed together. The commands are the same in either work flow if commits are squashed before a PR is opened, or if they are squashed later after review but before merge.

For the sake of being concrete, a work flow would look like this:

```
git clone ...
git checkout -b <a new branch>
... do stuff ..
git commit -m "comment about what was done"
... do stuff ..
git commit -m "more stuff was done"
... now tested and ready, let's do PR ...
git push origin <a new branch>
```

Then review happens, perhaps some more commits are added:
```
... do stuff .///
git commit -m "add some feedback review, maybe fix formatting or something."
git push origin < a new branch>
```

Now let's say the reviewers like it, and they say it is good to go. The ball is then in the submitter's court. Let's say for example there were 5 commits total as part of this feature, and the feature is logically one thing, let's say for example that thing is "Made the battle calculator multi-ithreaded"
```
https://github.com/DanVanAtta/linux_tools/blob/master/bin/git_squash 5
... Update git comment  "Made the battle calculator multi-ithreaded\n Details..."....
git push origin -f <a new branch>
```
Now submitter notifies that the commits have been squashed and it is ready for merge.

**Comment by DanVanAtta** [comment]
I propose this:
- I'll indicate explicitly in a PR if the "ball" should be passed back to me before merge so I can squash commits together.
- Otherwise if there are no specific instructions, a PR can be merged once the reviewers agree it is good to merge. 


(As an extended side note, the bar I like to define as ready to be merged is: "does this add more value than it does cost?" Both value and cost can be subjective. For consideration would be a well coded, not buggy, new feature that adds value. Perhaps it may not have all the bells and whistles or be perfect, but often it can be better to merge it in so integration can happen. Then following PRs can be very focused improvements. To make it complicated though there is a lot of gray area, and subjectivity involved, so I do not intend to propose that as necessarily a hard and fast rule, more a general guideline.)

**Comment by ron-murhammer** [comment]
@DanVanAtta So my 2 concerns around this are:
1. The final PR that is merged is different then what is initially reviewed so either there is extra overhead of a final review or we are making an assumption that the committer properly squashes commits and no other changes.
2. Is what we are gaining (less commits, cleaner history) worth the extra overhead of adding these steps into the process?

**Comment by DanVanAtta** [comment]
If you do interactive rebase there should be no chance for error. Otherwise squash should be pretty easy to do, after having done so over 50 times for this project alone, I've yet had any troubles to keep the before and after diff from being the same. The key perhaps is to ensure that before you squash you have a clean status and no modified files.

The cleaner history is a nice benefit, but a linear history is a *much* bigger benefit. If you have a pure squash rebase work flow that never contains any merge commits, then you have a linear history. If you squash rebase each feature branch into one commit before merge, then you will have an effectively linear history. If you do not squash commits, and have a commit with a time stamp that is before one that is already merged into master, than your history will not be linear.

For more details on the benefit of linear history, below are some good readings with some summaries. What it comes down to though is that with a linear history, once you check out a commit that breaks the build, if the commit does not, then you found the commit that broke your build. With a non-linear history, that is not true (so this is an example where Git deviates from SVN, in SVN you can always walk the commit history to find what broke a build, with a non-linear git history, that is not the case).

Here is an interesting read on the rebase work flow discussed in the context of Facebook best practices:  https://news.ycombinator.com/item?id=7648534 

Here are the atlassian people talking about this work flow: https://www.atlassian.com/git/articles/git-team-workflows-merge-or-rebase/ By squash rebasing the feature branches and merging in just one commit, we have an effectively linear history, and we gain most of the benefits. If we don't do the squash then the history is not linear.


This link gives diagrams of what a linear history would look like:http://code.tutsplus.com/tutorials/rewriting-history-with-git-rebase--cms-23191 
With that mental picture you can then see more easily why that is better than having forks in your history. When you merge two branches, you get interleaved commits on master. But if you check out any of the commits, you get the context of the commit at that time. So that means if you check out one commit after another in master according to the history, your commits are jumping between branches. In the SVN world if you checked out far enough back and found a working build, and went forward and found the first broken one, you found a build break. In git, with non-linear history, that is *NOT* true, and it can be a huge PITA to find build breaks in non-linear Git histories.


Some more discussion about how linear history makes build breaks easier to deal with: https://github.com/balanced/balanced-dashboard/issues/461

> "That's better than a cake. It is very easy to read what is going on in the history, you can know exactly what branches are merged. Also, there are only merged commits left on the master branch. This means, when something goes wrong in one branch, you can always rollbacks to previous workable merged commit in the master. Or you can always revert one merged branch which has some serious problems easily."


Interestingly here is an argument against doing rebase and squashes, basically a bit of an extreme on the other end of the spectrum. The comment on that blog though tears the non-ff (merge) argument apart, it's both interesting and relevant to the proposed workflow to squash rebase:  http://walkingthestack.blogspot.com/2012/05/why-you-should-use-git-merge-no-ff-when.html

**Comment by DanVanAtta** [comment]
Linear history can be a big benefit. 

We have a fast build that builds very often thanks to Travis - so we get less benefit than we could from linear history (if the build took hours to complete and there are many commits in that time, the build will be broken more often and finding build breaks will be more important).

So rather than responding to this, I'll sum things up and close this issue : )

The reading links above I think were pretty interesting, and I personally enjoy improving and getting a more efficient work flow.

So in conclusion:
- PRs will be sumitted in the form they are intended to be merged
- PRs will be rebased on to master as often as conveniently possible to keep branches fresh
- PRs will be optimized for ease-of-review rather than for git history browsing (ie: `git log` output).
